src/main/java/com/metamx/common/CompressionUtils.java

```diff
@@ -0,0 +1,326 @@
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Paths;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+public class CompressionUtils
+{
+  private static final Logger log = new Logger(CompressionUtils.class);
+  private static final int DEFAULT_RETRY_COUNT = 10;
+
+  public static final String GZ_SUFFIX = ".gz";
+  public static final String ZIP_SUFFIX = ".zip";
+
+  public static long zip(File directory, File outputZipFile) throws IOException
+  {
+    if (!isZip(outputZipFile.getName())) {
+      log.warn("No .zip suffix[%s], putting files from [%s] into it anyway.", outputZipFile, directory);
+    }
+
+    try (final FileOutputStream out = new FileOutputStream(outputZipFile)) {
+      return zip(directory, out);
+    }
+  }
+
+  public static long zip(File directory, OutputStream out) throws IOException
+  {
+    if (!directory.isDirectory()) {
+      throw new IOException(String.format("directory[%s] is not a directory", directory));
+    }
+    final File[] files = directory.listFiles();
+
+    long totalSize = 0;
+    try (final ZipOutputStream zipOut = new ZipOutputStream(out)) {
+      for (File file : files) {
+        log.info("Adding file[%s] with size[%,d].  Total size so far[%,d]", file, file.length(), totalSize);
+        if (file.length() >= Integer.MAX_VALUE) {
+          zipOut.finish();
+          throw new IOException(String.format("file[%s] too large [%,d]", file, file.length()));
+        }
+        zipOut.putNextEntry(new ZipEntry(file.getName()));
+        totalSize += Files.asByteSource(file).copyTo(zipOut);
+      }
+      zipOut.closeEntry();
+    }
+
+    return totalSize;
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource inputStreamSupplier,
+      final File outDir,
+      final Predicate<Throwable> shouldRetry,
+      Boolean cacheLocally
+  ) throws IOException
+  {
+    if (!cacheLocally) {
+      return unzip(inputStreamSupplier.openStream(), outDir);
+    }
+    final File tmpFile = File.createTempFile("compressionUtilZipCache", ZIP_SUFFIX);
+    try {
+      FileUtils.FileCopyResult copyResult = FileUtils.retryCopy(
+          inputStreamSupplier,
+          shouldRetry,
+          tmpFile,
+          DEFAULT_RETRY_COUNT
+      );
+      return unzip(tmpFile, outDir);
+    }
+    finally {
+      if (!tmpFile.delete()) {
+        log.warn("Could not delete zip cache at [%s]", tmpFile.getAbsolutePath());
+      }
+    }
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource inputStreamSupplier,
+      final File outDir,
+      Boolean cacheLocally
+  ) throws IOException
```

Comment:
the order of arguments for the version without retry predicate is inconsistent with the one with predicate, we should put the predicate last I think instead of adding it somewhere in the middle